src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -293,32 +293,63 @@ private bool FullyValidateAndReorgLocked()
+        /// </summary>
+        /// <param name="blockHashes">List of headers to download.</param>
+        /// <param name="onBlockDownloadedCallback">A callback to call when the block was downloaded.</param>
         private void DownloadBlocks(List<ChainedHeader> blockHashes, OnBlockDownloadedCallback onBlockDownloadedCallback)
         {
             this.logger.LogTrace("({0}.{1}:{2})", nameof(blockHashes), nameof(blockHashes.Count), blockHashes.Count);
 
-            BlockDownloadRequest request = new BlockDownloadRequest
-            {
-                BlocksToDownload = blockHashes,
-            };
+            List<BlockDownloadRequest> downloadRequests = new List<BlockDownloadRequest>();
+
+            BlockDownloadRequest request = new BlockDownloadRequest { BlocksToDownload = new List<ChainedHeader>() };
 
             lock (this.blockRequestedLock)
             {
-                foreach (ChainedHeader chainedHeader in request.BlocksToDownload)
+                ChainedHeader previousHeader = null;
+
+                foreach (ChainedHeader chainedHeader in blockHashes)
                 {
-                    if (!this.callbacksByBlocksRequesteHash.TryGetValue(chainedHeader.HashBlock, out List<OnBlockDownloadedCallback> listOfCallbacks))
+                    bool blockAlreadyAsked = false;
+                    if (this.callbacksByBlocksRequestedHash.TryGetValue(chainedHeader.HashBlock, out List<OnBlockDownloadedCallback> listOfCallbacks))
+                    {
+                        listOfCallbacks.Add(onBlockDownloadedCallback);
+                        blockAlreadyAsked = true;
+                    }
+                    else
+                    {
+                        this.callbacksByBlocksRequestedHash.Add(chainedHeader.HashBlock, new List<OnBlockDownloadedCallback>() { onBlockDownloadedCallback });
+                    }
+
+                    bool blockIsNotConsecutive = previousHeader != null && chainedHeader.Previous.HashBlock != previousHeader.HashBlock;
+
+                    if (blockIsNotConsecutive || blockAlreadyAsked)
                     {
-                        listOfCallbacks = new List<OnBlockDownloadedCallback>();
-                        this.callbacksByBlocksRequesteHash.Add(chainedHeader.HashBlock, listOfCallbacks);
+                        downloadRequests.Add(request);
+                        request = new BlockDownloadRequest { BlocksToDownload = new List<ChainedHeader>() };
+
+                        if (blockAlreadyAsked)
+                        {
+                            previousHeader = null;
+                            continue;
+                        }
                     }
 
-                    listOfCallbacks.Add(onBlockDownloadedCallback);
+                    previousHeader = chainedHeader;
+                    request.BlocksToDownload.Add(chainedHeader);
```

Comment:
How about the last batch?

How about two consecutive blocks that have been asked already?

Reply:
Ah yeah I will fix